server/src/main/scala/org/http4s/armeria/server/ArmeriaServerBuilder.scala

```diff
@@ -373,6 +381,12 @@ object ArmeriaServerBuilder {
       armeriaServerBuilder = defaultServerBuilder,
       socketAddress = defaults.IPv4SocketAddress,
       serviceErrorHandler = DefaultServiceErrorHandler,
-      banner = defaults.Banner)
+      banner = defaults.Banner,
+      dispatcher = dispatcher
+    )
   }
+
+  /** Returns a newly created [[org.http4s.armeria.server.ArmeriaServerBuilder]]. */
+  def apply[F[_] : Async]: Resource[F, ArmeriaServerBuilder[F]] =
```

Comment:
Is this what you were questioning having in a `Resource`?  If your builder builds a function of what to do with the `Dispatcher`, you can delay construction until you have a `Dispatcher`.  `JettyBuilder` does something similar to this with its `Mount`s.

Reply:
I looked at that kind of approach, but it seemed like it would require a much more radical refactoring because the `Dispatcher` is passed into the `ArmeriaHttp4sHandler` which is then passed into the `ServerBuilder` (`armeriaServerBuilder`), which is not (a priori) copyable. So e.g. if we add the `ArmeriaHttp4sHandler` during resource initialization then we end up with a `Resource` that's not actually resuable as it should be (because the mutable `ServerBuilder` ends up containing handlers that have a reference to the old `Dispatcher`). Maybe reworking the whole `ArmeriaServerBuilder` to not contain a persistent `ServerBuilder` is the right approach, and I'm happy to do that if you think it's a good way forward, but I wanted to try something less intrusive at first.

Reply:
My gut feeling is that's the right approach, but I defer entirely to @ikhoon.  I don't have any Armeria experience aside from dropping by PRs like these. :smile:

Reply:
I've check upstream's server builders such as `BlazeSErverBuilder`, `JettyBuilder`. They seem not to expose `Dispatcher` to users or take `Dispatcher` from users.
I'd like to give a similar API experience to users. 
How about lazily binding services when building a server?
```scala
// mountService should be called before building the server in `resource`
private var mountService: ArmeriaBuilder: Unit: _ => ()
def withHttpApp(prefix: String, service: HttpApp[F]): Self = {
  mountService = armeriaServerBuilder => {
    mountService(armeriaServerBuilder)
    armeriaServerBuilder.serviceUnder(prefix, ArmeriaHttp4sHandler(prefix, service, dispatcher))
  }
  this
}
```

Reply:
It would be nice to also apply to `withHttpService` to preserve the order of routing set by users.

Reply:
Yeah. I think it fits with `JettyBuilder` etc. to use a function rather than a mutable builder in our builder; this also means this builder can be immutable and use copy style. I've tried that approach just now and I think it's nicer (it let me simplify the TLS `Resource` case as well).
(I can't run the build locally due to some protobuf classpath confusion in intellij, but hopefully the intent is clear even if there are minor build errors left).